File: xtask/src/config_flash.rs
use anyhow::{bail, Context};
use serialport::{DataBits, FlowControl, Parity, StopBits};
use std::io::{Read, Write};
use std::time::{Duration, Instant};

/// Strip ANSI escape sequences from a string for reliable matching.
fn strip_ansi(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut chars = s.chars().peekable();
    while let Some(c) = chars.next() {
        if c == '\x1b' {
            // Skip the entire escape sequence
            if chars.peek() == Some(&'[') {
                chars.next();
                while let Some(&nc) = chars.peek() {
                    chars.next();
                    if nc.is_ascii_alphabetic() {
                        break;
                    }
                }
            }
        } else {
            out.push(c);
        }
    }
    out
}

pub fn run(args: &[String]) -> anyhow::Result<()> {
    let mut cli_port = None;
    let mut serial = None;
    let mut port_index = None;
    let mut no_reset = false;
    let mut parser = crate::util::Args::new(args);
    while parser.has_more() {
        match parser.peek().unwrap() {
            "--port" => cli_port = Some(parser.value("--port")?.to_string()),
            "--serial" => serial = Some(parser.value("--serial")?.to_string()),
            "--port-index" => port_index = parser.optional_parsed("--port-index")?,
            "--no-reset" => {
                no_reset = true;
                parser.skip();
            }
            _ => return Err(parser.unknown_arg()),
        }
    }

    let config = crate::config::load()?;
    let port_name = crate::util::resolve_port(
        cli_port.as_deref(),
        serial.as_deref(),
        port_index,
        Some(&config),
    )?;

    // Build config payload
    let mut payload = String::from("CONFIG_START\r\n");
    payload.push_str(&format!("wifi.ssid={}\r\n", config.wifi.ssid));
    payload.push_str(&format!("wifi.password={}\r\n", config.wifi.password));
    payload.push_str(&format!("mqtt.host={}\r\n", config.mqtt.host));
    payload.push_str(&format!("mqtt.port={}\r\n", config.mqtt.port));
    if !config.mqtt.user.is_empty() {
        payload.push_str(&format!("mqtt.user={}\r\n", config.mqtt.user));
    }
    if !config.mqtt.password.is_empty() {
        payload.push_str(&format!("mqtt.password={}\r\n", config.mqtt.password));
    }
    payload.push_str(&format!("device.id={}\r\n", config.device.id));
    payload.push_str("CONFIG_END\r\n");

    println!("Connecting to {}...", port_name);
    let mut port = serialport::new(&port_name, 115200)
        .data_bits(DataBits::Eight)
        .parity(Parity::None)
        .stop_bits(StopBits::One)
        .flow_control(FlowControl::None)
        .timeout(Duration::from_millis(100))
        .open()
        .with_context(|| format!("Failed to open {}", port_name))?;

    // Drain any stale data in the RX buffer
    let mut drain = [0u8; 4096];
    let _ = port.read(&mut drain);

    if !no_reset {
        // Reset ESP32 via DTR/RTS (standard auto-reset circuit on NodeMCU-style boards).
        println!("Resetting ESP32...");
        port.write_data_terminal_ready(false)
            .context("Failed to set DTR")?;
        port.write_request_to_send(true)
            .context("Failed to set RTS")?;
        std::thread::sleep(Duration::from_millis(100));
        port.write_request_to_send(false)
            .context("Failed to set RTS")?;
        println!("Waiting for ESP32 to boot...");
        std::thread::sleep(Duration::from_millis(5000));
        // Drain any boot output
        let _ = port.read(&mut drain);
    }

    println!("Sending config...");
    // Send line-by-line with 50ms delays to avoid overflowing the ESP32's
    // 128-byte UART RX FIFO. The total config payload (~180 bytes) exceeds
    // the FIFO size, so sending all at once causes silent data loss.
    for line in payload.lines() {
        port.write_all(line.as_bytes())
            .context("Failed to write config line")?;
        port.write_all(b"\r\n")
            .context("Failed to write line ending")?;
        port.flush().context("Failed to flush")?;
        std::thread::sleep(Duration::from_millis(150));
    }

    // Wait for CONFIG_OK or CONFIG_ERROR
    let resp_deadline = Instant::now() + Duration::from_secs(10);
    let mut buf = [0u8; 4096];
    let mut response = String::new();
    while Instant::now() < resp_deadline {
        match port.read(&mut buf) {
            Ok(0) => {}
            Ok(n) => {
                response.push_str(&String::from_utf8_lossy(&buf[..n]));
                let clean = strip_ansi(&response);
                if clean.contains("CONFIG_OK") {
                    println!("Config written successfully!");
                    return Ok(());
                }
                if clean.contains("CONFIG_ERROR") {
                    bail!("ESP32 config error: {}", clean.trim());
                }
            }
            Err(ref e) if e.kind() == std::io::ErrorKind::TimedOut => {}
            Err(e) => bail!("Serial read error: {}", e),
        }
    }
    bail!(
        "No CONFIG_OK response within 10s. Output:\n{}",
        strip_ansi(&response)
    );
}

File: xtask/scripts/config_flash.py
"""Config flash utility for Launa ESP32.

Reads a config payload file and sends it over serial to an ESP32
waiting for configuration via the serial config protocol.

Usage: python scripts/config_flash.py <port_name> <config_file>
"""

import serial
import time
import sys


def main():
    if len(sys.argv) < 3:
        print("Usage: config_flash.py <port_name> <config_file>", file=sys.stderr)
        sys.exit(1)

    port_name = sys.argv[1]
    config_file = sys.argv[2]

    with open(config_file, "r") as f:
        config_lines = f.read()

    print(f"Opening {port_name}...", file=sys.stderr)
    port = serial.Serial(port_name, 115200, timeout=1)

    # Toggle DTR to reset the ESP32, then catch the config window on boot
    print("Resetting ESP32 via DTR...", file=sys.stderr)
    port.dtr = False
    time.sleep(0.1)
    port.dtr = True
    time.sleep(0.1)
    port.dtr = False

    # Wait for ESP32 ready signal after reboot
    start = time.time()
    ready = False
    all_output = ""
    while time.time() - start < 35:
        data = port.read(4096)
        if data:
            text = data.decode("utf-8", errors="replace")
            all_output += text
            sys.stderr.write(text)
            sys.stderr.flush()
            if "Waiting for serial config" in all_output:
                ready = True
                break

    if not ready:
        print(
            f"ERROR: ESP32 not ready within 35s. Output so far: {all_output}",
            file=sys.stderr,
        )
        port.close()
        sys.exit(1)

    print("ESP32 ready, sending config...", file=sys.stderr)
    time.sleep(0.1)

    # Send each config line with CRLF line endings
    for line in config_lines.strip().split("\n"):
        port.write((line + "\r\n").encode("utf-8"))
        print(f"  Sent: {line}", file=sys.stderr)
    port.flush()

    # Wait for response
    start = time.time()
    response = b""
    while time.time() - start < 10:
        data = port.read(4096)
        if data:
            response += data
            sys.stderr.write(data.decode("utf-8", errors="replace"))
            sys.stderr.flush()
            if b"CONFIG_OK" in response or b"CONFIG_ERROR" in response:
                break

    port.close()

    if b"CONFIG_OK" in response:
        print("CONFIG_OK")
    elif b"CONFIG_ERROR" in response:
        idx = response.find(b"CONFIG_ERROR:")
        error = response[idx + len(b"CONFIG_ERROR:") :].decode("utf-8", errors="replace").strip()
        print(f"CONFIG_ERROR: {error}")
    else:
        decoded = response.decode("utf-8", errors="replace").strip()
        print(f"NO_RESPONSE: {decoded}")


if __name__ == "__main__":
    main()
